lib/web/desktop_playback.go

```diff
@@ -0,0 +1,246 @@
+	defer pp.log.Debug("playbackPlayer.ReceiveActions returned")
+	defer pp.close(cancel)
+
+	for {
+		action := playbackAction{}
+		// Hangs until there is data to be received, or until an error.
+		err := websocket.JSON.Receive(pp.ws, &action)
+		if err != nil {
+			// We expect net.ErrClosed if the websocket is closed by another
+			// goroutine and io.EOF if the websocket is closed by the browser
+			// while websocket.JSON.Receive() is hanging.
+			if !errors.Is(err, net.ErrClosed) && !errors.Is(err, io.EOF) {
+				pp.log.WithError(err).Error("error reading from websocket")
+			}
+			return
+		}
+		pp.log.Debugf("recieved playback action: %+v", action)
+		switch action.Action {
+		case actionPlayPause:
+			pp.togglePlaying()
+		default:
+			pp.log.Errorf("received unknown action: %v", action.Action)
+			return
+		}
+	}
+}
+
+// streamSessionEvents streams the session's events as playback events over the websocket.
+func (pp *playbackPlayer) streamSessionEvents(ctx context.Context, cancel context.CancelFunc) {
+	defer pp.log.Debug("playbackPlayer.StreamSessionEvents returned")
+	defer pp.close(cancel)
+
+	var lastDelay int64
+	eventsC, errC := pp.clt.StreamSessionEvents(ctx, session.ID(pp.sID), 0)
+	for {
+		pp.waitWhilePaused()
+
+		select {
+		case err := <-errC:
+			if !errors.Is(err, context.Canceled) {
+				pp.log.WithError(err).Errorf("streaming session %v", pp.sID)
+			}
+
+			return
+		case evt := <-eventsC:
+			if evt == nil {
+				pp.log.Debug("reached end of playback")
+
+				if _, err := pp.ws.Write([]byte(`{"message":"end"}`)); err != nil {
+					pp.log.WithError(err).Error("failed to write \"end\" message over websocket")
+				}
+
+				// deferred ps.Close() will set ps.playState = playStateFinished for us
+				return
+			}
+			switch e := evt.(type) {
+			case *apievents.DesktopRecording:
+				if e.DelayMilliseconds > lastDelay {
+					time.Sleep(time.Duration(e.DelayMilliseconds-lastDelay) * time.Millisecond)
+					lastDelay = e.DelayMilliseconds
+				}
+				msg, err := utils.FastMarshal(e)
+				if err != nil {
+					pp.log.WithError(err).Errorf("failed to marshal DesktopRecording event into JSON: %v", e)
+				}
+				if _, err := pp.ws.Write(msg); err != nil {
+					// We expect net.ErrClosed to arise when another goroutine returns before
+					// this one or the browser window is closed, both of which cause the websocket to close.
+					if !errors.Is(err, net.ErrClosed) {
+						pp.log.WithError(err).Error("failed to write DesktopRecording event over websocket")
+					}
+					return
+				}
+			default:
+				pp.log.Warnf("session %v contains unexpected event type %T", pp.sID, evt)
+			}
+		}
+	}
+}
+
+// Play kicks off goroutines for receiving actions
+// and playing back the session over the websocket,
+// and then hangs until the websocket is closed.
+func (pp *playbackPlayer) Play(ctx context.Context) {
+	defer pp.log.Debug("playbackPlayer.Play returned")
+	ppCtx, cancel := context.WithCancel(ctx)
+	defer pp.close(cancel)
```

Comment:
I found passing an external cancel like this a bit strange, specially when reading the methods that take CancelFunc as an argument.

If we are not allowing players to be reused (which I think is the case), then maybe consider having an internal cancel that the funcs can access?

Reply:
This came about because I'm not a fan of holding on to `context.Context` (or its associated cancel funcs) as struct members and asked Isaiah to see if we could pass them as arguments instead.

In this case, I guess we could capture the CancelFunc as a member of the player struct, but my newfound Rust experience is telling me this is a bad idea. The context is only valid for the duration of `Play`, so we shouldn't stash the associated cancel somewhere that could outlive `Play`.

Have any ideas?

Reply:
If we don't allow more than a single invocation of Play, which seems to be the intent, then my initial idea would be to nest playCtx/playCancel fields inside playbackPlayer, initialized on new.

In general I don't like concurrency primitives passed around - funcs with channels, locks or cancels as params. I'd either handle concurrency where the goroutines are launched (ie, sync primitives live in a single method) or nest attributes. That said, what you have definitely works and it's all private, so feel free to push back if you don't think this is worth pursuing.

Reply:
I don't love either approach, to be honest. Right now, we don't allow more than one invocation of `Play()`. But that might change. 

I'd prefer that we continue to pass the context to play, as that is the long running operation that must support cancelation, and I can't really think of a better way to maintain that invariant without passing around a shared cancelation signal.